### DISCUSSION TITLE
/zh-cn/blog/posts/when_and_where_to_use_macro/

### DISCUSSION THREAD

**Discussion by Reverier-Xu** [Reverier-Xu/discussions#130] [Announcements]
https://blog.woooo.tech/posts/when_and_where_to_use_macro/

**Comment by hu4i** [comment]
Evaluate就翻译成求值吧，“评估”有点感觉是像在计算可行性、代价之类的。这里evaluate就是求表达式的值。


Rust的宏是语法树的一部分([Macros in the AST](https://danielkeep.github.io/tlborm/book/mbe-syn-macros-in-the-ast.html))，这也就表示其展开前必须作为一个合法的语法树的节点。所以其做不了C语言诸如`#define IF(x) if(x){`这样迷惑操作，但是因为其在语法分析过程中展开，能够让我们捕获语法单元，甚至是注释，这样可以让我们包装结构体和函数等实体，从而设计出很多感觉是魔法的特性。比C的宏更适合隐藏细节，对使用你的代码的人更友好。

不过有些东西你点进去确实是魔法(`/* compiler built-in */`)：
```rust
#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_builtin_macro]
#[macro_export]
#[cfg_attr(not(test), rustc_diagnostic_item = "stringify_macro")]
macro_rules! stringify {
    ($($t:tt)*) => {
        /* compiler built-in */
    };
}
```

C语言的宏和Rust的宏都不能滥用，但是Rust的宏不会破坏语法结构，导致出现`#define IF(x) if(x){`之类引起的Bug。

你可以在Rust里设计你自己的语言：[Ook!](https://danielkeep.github.io/tlborm/book/aeg-ook.html)，对标Brainfuck。希望不要出现在CTF逆向里（

**Comment by Reverier-Xu** [comment]
> @hu4i
> Evaluate就翻译成求值吧，“评估”有点感觉是像在计算可行性、代价之类的。这里evaluate就是求表达式的值。
> 
> 
> Rust的宏是语法树的一部分([Macros in the AST](https://danielkeep.github.io/tlborm/book/mbe-syn-macros-in-the-ast.html))，这也就表示其展开前必须作为一个合法的语法树的节点。所以其做不了C语言诸如`#define IF(x) if(x){`这样迷惑操作，但是因为其在语法分析过程中展开，能够让我们捕获语法单元，甚至是注释，这样可以让我们包装结构体和函数等实体，从而设计出很多感觉是魔法的特性。比C的宏更适合隐藏细节，对使用你的代码的人更友好。
> 
> 不过有些东西你点进去确实是魔法(`/* compiler built-in */`)：
> ```rust
> #[stable(feature = "rust1", since = "1.0.0")]
> #[rustc_builtin_macro]
> #[macro_export]
> #[cfg_attr(not(test), rustc_diagnostic_item = "stringify_macro")]
> macro_rules! stringify {
>     ($($t:tt)*) => {
>         /* compiler built-in */
>     };
> }
> ```
> 
> C语言的宏和Rust的宏都不能滥用，但是Rust的宏不会破坏语法结构，导致出现`#define IF(x) if(x){`之类引起的Bug。
> 
> 你可以在Rust里设计你自己的语言：[Ook!](https://danielkeep.github.io/tlborm/book/aeg-ook.html)，对标Brainfuck。希望不要出现在CTF逆向里（

orz 修改了, 出题是个好主意 (XD, 明年比赛拿去欺负学弟 (bushi

**Comment by framist** [comment]
“求值”感觉也可以这么解释：`macro` 是正则序求值，`function` 是应用序求值。
